### ISSUE TITLE
MeetingAttendanceReport returns fordidden code with application permissions (with application access policy)

### ISSUE THREAD

**Issue by alsyfadis** [microsoftgraph/microsoft-graph-docs#12670]
Hi, 

I just want to know if it 's possible to get the MeetingAttendanceReport with application permissions.
Right now, it's possible to create and get online meetings with application permissions and an application access policy. 

The new feature MeetingAttendanceReport is available with delegated permissions but returns 403 error (code Forbidden), even if the OnlineMeetings.ReadWrite.All application permission is set in my token.

Thanks for your help !

Antoine
---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9a9593d8-ed60-527f-c549-b22dce5adbd9
* Version Independent ID: cdf36bed-fd4f-8be7-918c-6cdf74258676
* Content: [Get onlineMeeting - Microsoft Graph beta](https://docs.microsoft.com/en-us/graph/api/onlinemeeting-get?view=graph-rest-beta&tabs=http)
* Content Source: [api-reference/beta/api/onlinemeeting-get.md](https://github.com/microsoftgraph/microsoft-graph-docs/blob/main/api-reference/beta/api/onlinemeeting-get.md)
* Product: **cloud-communications**
* Technology: **microsoft-graph**
* GitHub Login: @jsandoval-msft
* Microsoft Alias: **MSGraphDocsVteam**

**Comment by ffactory-ofcl** [comment]
I'm also having this issue. 
I understand that only the organizer is allowed to get attendance but as the application access policy already allows creating meetings on behalf of a user, it should also allow retrieving the report on behalf of it.

**Comment by enkelmedia** [comment]
Same issue for me as well.

There is a thread on the community forum here as well: https://docs.microsoft.com/en-us/answers/questions/385445/onlinemeeting-and-meetingattendancereport-returns.html

**Comment by ricipi** [comment]
Same issue....It's working from Microsoft Graph Explorer but if i try with a flow in Power Automate i got "Forbidden" as response message.

Here the full response:
{
    "statusCode": 403,
    "headers": {
        "Transfer-Encoding": "chunked",
        "Vary": "Accept-Encoding",
        "Strict-Transport-Security": "max-age=31536000",
        "request-id": "99fe81b9-c5fa-4964-9542-e00555a6a864",
        "client-request-id": "99fe81b9-c5fa-4964-9542-e00555a6a864",
        "x-ms-ags-diagnostic": "{\"ServerInfo\":{\"DataCenter\":\"West Europe\",\"Slice\":\"E\",\"Ring\":\"5\",\"ScaleUnit\":\"002\",\"RoleInstance\":\"AM2PEPF000054F2\"}}",
        "scenario-id": "5eccb032-9787-4eae-8399-732462b19eca",
        "Date": "Tue, 25 May 2021 14:46:18 GMT",
        "Content-Type": "application/json",
        "Content-Length": "207"
    },
    "body": {
        "error": {
            "code": "Forbidden",
            "message": "Forbidden",
            "innerError": {
                "date": "2021-05-25T14:46:19",
                "request-id": "99fe81b9-c5fa-4964-9542-e00555a6a864",
                "client-request-id": "99fe81b9-c5fa-4964-9542-e00555a6a864"
            }
        }
    }
}

**Comment by frankpeng7** [comment]
Hi guys, thank you for your interest in the new API :) Let me reply to the comments in this thread in a batch:

@alsyfadis @ffactory-ofcl: For some compliance reason, `/meetingAttendeeReport` path is only supporting delegated token at this moment. We're actively working on supporting app permission. We will update documentation and changelog once it's done.

@enkelmedia @ricipi: the issue that delegated token is not working outside of Graph Explorer has been fixed. Please expect by the end of next week for the change to roll out.

**Comment by ffactory-ofcl** [comment]
@frankpeng7 thanks for letting us know!

**Comment by alsyfadis** [comment]
@frankpeng7 Thanks you guys for working on this point !

**Comment by eprigorodov** [comment]
What can be also related is that Graph can return the same "Forbidden" error even when using delegated tokens. In our case an example request to `/beta/me/onlineMeetings/{meetingId}/meetingAttendanceReport` worked perfectly in the Graph Explorer and failed in the Python code with MSAL.

The difference was found in the `"aud"` claim: is was equal to `"00000003-0000-0000-c000-000000000000"` in the Graph Explorer and to `"https://graph.microsoft.com"` in tokens acquired from the Python code.

That was caused by different scopes: the Python code requested `['https://graph.microsoft.com/.default']` and GraphExplorer requests `['User.Read']`.

Changing requested scopes to `['OnlineMeetings.Read']` fixed the issue: Identity Platform returned a token with `"aud": "00000003-0000-0000-c000-000000000000"` and then it was accepted by "/meetingAttendanceReport" call.

Question: where one can find documentation of this behaviour?

**Comment by ffactory-ofcl** [comment]
@eprigorodov I strongly suggest you open a new issue about this or find an existing one. Your experience doesn't really relate to this problem.

**Comment by frankpeng7** [comment]
@eprigorodov Glad you found out where the issue is :) This is actually the workaround I mentioned in an earlier reply but you won't need to change the scope when requesting an auth token once the fix from our side gets deployed by the end of next week.

**Comment by frankpeng7** [comment]
Closing as this should've been fixed by now. Feel free to reopen if the issue persists.

**Comment by mduvaise** [comment]
Hi @frankpeng7  this issue still exists when /meetingAttendanceReport API is called with application permissions

**Comment by frankpeng7** [comment]
> Hi guys, thank you for your interest in trying out the new API. Let me reply to the comments in this thread in a batch:
> 
> @alsyfadis @ffactory-ofcl: For some compliance reason, `/meetingAttendeeReport` path is only supporting delegated token at this moment. We're actively working on supporting app permission. We will update documentation and changelog once it's done.
> 
> @enkelmedia @ricipi: the issue that delegated token is not working outside of Graph Explorer has been fixed. Please expect by the end of next week for the change to roll out. Feel free to message me if you need a workaround before that time.

@mduvaise please see my previous reply. We don't support `/meetingAttendeeReport` with app permission yet (but it's coming soon)

**Comment by enkelmedia** [comment]
@frankpeng7 I just tested with a token that I got by using grant_type=password, this used to return forbidden but works as expected now. Great work!

Reg. app permissions, this is when I create a token using the appID and appSecret right? Just so that we're talking about the same thing. You're saying that your actively working on this, and the end goal is to support this as well? I know it's hard but do you have any timeframe? Are we talking weeks, months?

All the best!

**Comment by HamzaJosh** [comment]
> What can be also related is that Graph can return the same "Forbidden" error even when using delegated tokens. In our case an example request to `/beta/me/onlineMeetings/{meetingId}/meetingAttendanceReport` worked perfectly in the Graph Explorer and failed in the Python code with MSAL.
> 
> The difference was found in the `"aud"` claim: is was equal to `"00000003-0000-0000-c000-000000000000"` in the Graph Explorer and to `"https://graph.microsoft.com"` in tokens acquired from the Python code.
> 
> That was caused by different scopes: the Python code requested `['https://graph.microsoft.com/.default']` and GraphExplorer requests `['User.Read']`.
> 
> Changing requested scopes to `['OnlineMeetings.Read']` fixed the issue: Identity Platform returned a token with `"aud": "00000003-0000-0000-c000-000000000000"` and then it was accepted by "/meetingAttendanceReport" call.
> 
> Question: where one can find documentation of this behaviour?

I tried this just by changing the scope from  scope = ['https://graph.microsoft.com/.default'] to scope = ['OnlineMeetings.Read']

Am I doing something wrong? How did you get this to work? PLEASE HELP

**Comment by frankpeng7** [comment]
Hi @HamzaJosh The fix has been rolled out. You don't need any additional changes to make this work.

**Comment by HamzaJosh** [comment]
> Hi @HamzaJosh The fix has been rolled out. You don't need any additional changes to make this work.
I am still getting this error today, when will I see it working?

**Comment by frankpeng7** [comment]
> > Hi @HamzaJosh The fix has been rolled out. You don't need any additional changes to make this work.
> > I am still getting this error today, when will I see it working?

It's already in effect . I think other users in this thread have confirmed it's working. Did you follow this guide to get delegated token? https://docs.microsoft.com/en-us/graph/auth-v2-user?context=graph%2Fapi%2Fbeta&view=graph-rest-beta

**Comment by HamzaJosh** [comment]
@frankpeng7  I read thru the article, it refers to a different scenario than mine.  My scenario is like this. I have registered an app in Azure Portal, given the app permissions to read onlinemeetings and calendar. I am using databricks writing python code to access the Graph API. I can read everything, users meetings etc. when I use the same access token to read Attendance meeting report, I get response forbidden.
Here's the code I am using to get access token. Please tell me what I am doing wrong

app = msal.ConfidentialClientApplication(clientID, authority=authority, client_credential = clientSecret)
scope = ['https://graph.microsoft.com/.default']
accessToken = app.acquire_token_silent(scope, account=None)

I can use above accesstoken and read this url

"https://graph.microsoft.com/v1.0/users/" + hamza + "/onlineMeetings?$filter=JoinWebUrl%20eq%20'" + enc + "'"

but when I try 

arurl = "https://graph.microsoft.com/beta/users/" + hamza + "/onlineMeetings/" + meetingid + "/meetingAttendanceReport"
it returns forbidden, Please advise what I should do?

**Comment by frankpeng7** [comment]
@HamzaJosh looks like you're using an application token right? If so please see my previous reply

> > Hi guys, thank you for your interest in trying out the new API. Let me reply to the comments in this thread in a batch:
> > @alsyfadis @ffactory-ofcl: For some compliance reason, `/meetingAttendeeReport` path is only supporting delegated token at this moment. We're actively working on supporting app permission. We will update documentation and changelog once it's done.
> > @enkelmedia @ricipi: the issue that delegated token is not working outside of Graph Explorer has been fixed. Please expect by the end of next week for the change to roll out. Feel free to message me if you need a workaround before that time.
> 
> @mduvaise please see my previous reply. We don't support `/meetingAttendeeReport` with app permission yet (but it's coming soon)

**Comment by HamzaJosh** [comment]
As you must have figured, I am newbie, can anyone tell me how to get delegated token using python in DataBricks?

**Comment by ffactory-ofcl** [comment]
That's a completely different issue than the one discussed here. See
the microsoft
graph docs <https://docs.microsoft.com/en-us/graph/permissions-reference>,
search on stackoverflow or find another github issue regarding your
question.

**Comment by eprigorodov** [comment]
@HamzaJosh, your code uses `msal.ConfidentialClientApplication()` and supplies `account=None` to `acquire_token_silent()`, which means obtaining *application token*.

As it was said here, application token does not work for the path `/meetingAttendanceReport` yet.

The workaround is to make the app work on behalf of user, i.e. impersonate meeting organizer account. Here is a minimal example of making that:

Prerequisites: turn on "Allow public client flows" in the app registration / Authentication blade.

```
msal_app = msal.PublicClientApplication(
    <application_id>,
    authority='https://login.microsoftonline.com/{}'.format(<tenant_id>)
)

accounts = msal_app.get_accounts(username=<username>)

if accounts:
    tokens = msal_app.acquire_token_silent(scopes=scopes, account=accounts[0])

if not accounts or not tokens:
    tokens = msal_app.acquire_token_by_username_password(
        username, password, scopes=['OnlineMeetings.Read']
    )
    access_token = tokens['access_token']
```

Caveats: storing plain text passwords is a dangerous practice and using `acquire_token_by_username_password()` is not recommended. It might be safer to wait for the fix for application token.

With regard to being a newbie, -- since the API is beta, it is natural to meet a lot of such obstacles, which require certain debugging skill. Programming by collecting pieces of code won't produce a good result.

**Comment by enkelmedia** [comment]
After playing around with this, 

Even when the meeting is "in progress" I get a attendance report back, but this is empty (as the docs says). The problem is that I'm not sure how to distinguish between a meeting that is in progress and a meeting that did not have any attendees.

The integration that we do will create a meeting and present the link to our users, but there is no guarantee that they really enter the meeting. And as far as I can see there is nothing in the data about the meeting that indicates if the meeting is really "going on" or if its closed? 

I guess that either there should be some indicator on the attendanceReport or on the meeting it self to that it's possible to know why the report is empty in this scenario.

**Comment by eprigorodov** [comment]
@enkelmedia, we played with about 100 online meetings, and it seems like meeting is considered to be in progress until the time stamped in its [property `endDateTime`](https://docs.microsoft.com/en-us/graph/api/resources/onlinemeeting?view=graph-rest-1.0#properties).

**Comment by enkelmedia** [comment]
@eprigorodov Thank you so much! I'm actually trying this out right now as well, I'll let you know. But as far as I understand, the actual video-meeting could still continue after the end time, Teams would not just kill the meeting right? In that case, if the host does not "End" the meeting, when would teams consider the meeting "Ended" and populate the report?

I guess I might just have to pull the report a couple of times for x hours and give up if no attendance is recorded say 12 hours after the meeting was scheduled to end. But still missing explicit information I don't like to guess =D

**Comment by eprigorodov** [comment]
@enkelmedia, sorry, we are also still debugging that and didn't find exact rules yet. Just one important observation: if the meeting is finished /in all senses/ and the attendance report is available, the meeting remains available for joining again. If any of attendees joins again then the first attendance report get overwritten with the new version that only contains these late-comers.

So yes, the approach with polling seems to be the most reliable.

**Comment by enkelmedia** [comment]
@eprigorodov Thank you so much for sharing your experience! When I tested now, the report will not be available if anyone is still in the meeting after the endTime is passed, as soon as everyone left the meeting the report will be available. But your point req. the overwrite is really valid for some use cases. 

Thanks again!
